### ISSUE TITLE
Updating relationships: an API MAY, SHOULD, or MUST support this?

### ISSUE THREAD

**Issue by adam-lynch** [json-api/json-api#1172]
From the specification:

> ### Updating Relationships

>Although relationships can be modified along with resources (as described above), JSON API also supports updating of relationships independently at URLs from relationship links.

"supports" is a bit unclear there.

---

Just a little bit after that in the specification:

> #### Updating To-One Relationships

>A server **MUST** respond to `PATCH` requests to a URL from a to-one relationship link as described below.

Which one of the following is correct here?

- "A server **MUST** respond to `PATCH` requests to a URL from a to-one relationship link..." if they (relationship links) exist, i.e. you choose to support them.
- "A server **MUST** respond to `PATCH` requests to a URL from a to-one relationship link..." which must exist (for every relationship of every entity).

**Comment by beauby** [comment]
I have always interpreted it as the former (i.e. "if the relationship link exists"). I agree the spec could be clearer on relationship endpoints though.

**Comment by ethanresnick** [comment]
I agree the spec could maybe be clearer, but @beauby's interpretation is correct here. Relationship links are optional (as noted in the definition of [the relationships object](http://jsonapi.org/format/#document-resource-object-relationships), which only must contain *"one of ..."*), and obviously the server can't/needn't support PATCHing at endpoints that don't exist.

**Comment by cmeeren** [comment]
If the server supports updating a relationship via a relationship link, does it also have to support updating the relationship through the resource endpoint? Or is it possible to *only* support updating relationships through relationship endpoints?

**Comment by jelhan** [comment]
> If the server supports updating a relationship via a relationship link, does it also have to support updating the relationship through the resource endpoint? Or is it possible to _only_ support updating relationships through relationship endpoints?

The specification explicitly allows a server to reject a request to update a resource's to-many relationship field:

> A server MAY reject an attempt to do a full replacement of a to-many relationship. In such a case, the server MUST reject the entire update, and return a 403 Forbidden response.

It does not mention the to-one case. I agree that it would be helpful to clarify.

**Comment by lindyhopchris** [comment]
@jelhan I find the insistence on rejecting it with a `403 Forbidden` response a bit restrictive. A `405 Method Not Allowed` would seem suitable for rejecting particular relationship requests, especially as that's what frameworks tend to use if the method has not been registered as a route.

**Comment by freddrake** [comment]
I generally reserve ``403 Forbidden`` for the case where someone is allowed to perform the operation, but not the requestor, though they're allowed to see the relevant resources.  Agree that ``405 Method Not Allowed`` is best where the method is not allowed is unsupported.

**Comment by jelhan** [comment]
I agree with both of you in general. But in this specific case a `PATCH` request to update the resource is allowed. The server may reject it only if it includes a to-many relationship field. Personally I would have preferred if it's handled as any other validation error - responding with `400 Bad Request` or `422 Unprocessable Entity`. But I'm not sure if this can be changed without breaking backward compatibility.
